randovania/games/am2r/gui/preset_settings/am2r_teleporters_tab.py

```diff
@@ -0,0 +1,144 @@
+
+    def _create_source_teleporters(self) -> None:
+        row = 0
+        region_list = self.game_description.region_list
+
+        locations = TeleporterList.nodes_list(self.game_enum)
+        node_identifiers: dict[NodeIdentifier, Area] = {
+            loc: region_list.area_by_area_location(loc.area_identifier) for loc in locations
+        }
+        checks: dict[NodeIdentifier, QtWidgets.QCheckBox] = {
+            loc: self._create_check_for_source_teleporters(loc) for loc in locations
+        }
+
+        self._teleporters_source_for_location = copy.copy(checks)
+        self._teleporters_source_destination: dict[NodeIdentifier, NodeIdentifier | None] = {}
+
+        for location in sorted(locations, key=lambda loc: (0, checks[loc].text())):
+            if location not in checks:
+                continue
+
+            self.teleporters_source_layout.addWidget(checks.pop(location), row, 1)
+
+            other_locations = [
+                node.default_connection
+                for node in node_identifiers[location].nodes
+                if isinstance(node, DockNode)
+                and node.dock_type in self.teleporter_types
+                and region_list.identifier_for_node(node) == location
+            ]
+            assert len(other_locations) == 1
+            teleporter_in_target = typing.cast(DockNode, region_list.node_by_identifier(location)).default_connection
+
+            self._teleporters_source_destination[location] = None
+
+            if teleporter_in_target:
+                other_loc = teleporter_in_target
+
+                if other_loc in checks:
+                    self.teleporters_source_layout.addWidget(checks.pop(other_loc), row, 2)
+                    self._teleporters_source_destination[location] = other_loc
+
+            row += 1
+
+    def on_preset_changed(self, preset: Preset) -> None:
+        config = preset.configuration
+        assert isinstance(config, AM2RConfiguration)
+        config_teleporters: TeleporterConfiguration = config.teleporters
+
+        descriptions = [
+            "<p>Controls where each teleporter pipe connects to.</p>",
+            f" {self.teleporter_mode_to_description[config_teleporters.mode]}</p>",
+        ]
+        self.teleporters_description_label.setText("".join(descriptions))
+
+        signal_handling.set_combo_with_value(self.teleporters_combo, config_teleporters.mode)
+        can_shuffle_source = config_teleporters.mode not in (TeleporterShuffleMode.VANILLA,)
+        can_shuffle_target = config_teleporters.mode not in (
+            TeleporterShuffleMode.VANILLA,
+            TeleporterShuffleMode.TWO_WAY_RANDOMIZED,
+            TeleporterShuffleMode.TWO_WAY_UNCHECKED,
+        )
+
+        for origin, destination in self._teleporters_source_destination.items():
+            origin_check = self._teleporters_source_for_location[origin]
+            dest_check = self._teleporters_source_for_location.get(destination)
```

Comment:
mypy failure.
Is there a use case where this really can be `None`?
L87 should always be true by the check I wrote the question? Is this for teleporter things leading to none teleporter things? Like a teleporter to the end credits?

Reply:
Yeah. I technically don't have any cases *right now* where that can happen. But I also don't wanna fix things in the future in case I introduce things like that, so I'll just fix the mypy error instead.